evaluation/utils.py

```diff
@@ -0,0 +1,83 @@
+import numpy as np
+import xml.etree.ElementTree as ET
+import yaml
+
+# dist is a map of string->list[];
+# return a list of tuple (list[], list[]),
+# xml namespace & a list of values.
+def param_sweep_space(ps_dist):
+    ps_space = []
+    for name in ps_dist:
+        val_list = ps_dist[name]
+        if val_list is None or len(val_list) == 0:
+            raise Exception('parameter space should not be empty.')
+        
+        xml_level = list(name.strip().split('.'))
+
+        # linear scale range
+        if (val_list[0] == '.range'):
+            if len(val_list) not in (3, 4):
+                raise Exception('incorrect argument number for linear scale')
+            if (val_list == 3):
+                val_list = np.arange(val_list[1], val_list[2], 1).tolist()
+            else:
+                val_list = np.arange(val_list[1], val_list[2], val_list[3]).tolist()
+
+        ps_space.append((xml_level, val_list))
+
+    return ps_space
+
+# inject list of param to certain xml file, given that file path name.
+# return an xml etree structure.
+def inject_param_xml(file_path, params):
+    conf_etree = ET.parse(file_path)
+    root = conf_etree.getroot()
+    for name_level, val in params:
+        cursor = root
+        # traverse xml name levels.
+        for key in name_level:
+            cursor = cursor.find(key)
+            if cursor is None:
+                break
+
+        if cursor is None:
+            print('Warning: fail to inject parameter in conf file,', name_level)
+            continue
+
+        cursor.text = str(val)
+
+    conf_etree.write(file_path)
+
+# recursive calling routine on parameter space.
+# f is a callback function with closure, it should deside
+# what to do with a permutation list of parameters.
+def parameter_sweep(ps_space, f, closure=dict()):
+    parameter_sweep_r(ps_space, [], f, closure)
+
+def parameter_sweep_r(ps_space, cur_params, f, closure):
+    if len(cur_params) == len(ps_space):
+        f(cur_params, closure)
+        return
+    
+    name_level, val_list = ps_space[len(cur_params)]
+    for val in val_list:
+        cur_params.append((name_level, val))
+        parameter_sweep_r(ps_space, cur_params, f, closure)
```

Comment:
Additionally, consider whether recursion is the right approach. Python has no tail-call optimization which means that your code will stop working after recursion depth 1000 (`import sys; sys.getrecursionlimit()`).

Try to convert your sweeping to use a loop instead.

Reply:
I am not sure which one is better, because making this non-recursive has two caveats:
- Making the traverse non-recursive leverages some tricky stack logic, which will lower code readability.
- Unless one makes 1000+ parameters to tune, the calling stack would never exceed the limit. Normally we have lower than 10 so the maximum layer would be around 10.